drivers/soundwire/bus.c

```diff
@@ -1013,6 +1019,10 @@ static int sdw_handle_slave_alerts(struct sdw_slave *slave)
 	if (count == SDW_READ_INTR_CLEAR_RETRY)
 		dev_warn(slave->bus->dev, "Reached MAX_RETRY on alert read\n");
 
+io_err:
+	pm_runtime_mark_last_busy(&slave->dev);
+	pm_runtime_put(&slave->dev);
```

Comment:
The commit message doesn't really describe the issue and I am having a hard time convincing myself.

An alert can only come when the bus is operational and the device is enumerated with a non-zero device number.
So in theory read/write access to registers should not be an issue.

It would really help if you could describe in more details what the root cause is, so that we can justify this change.

Reply:
@plbossart   For rt711, the suspend function only change regmap to cache only. And in this case, slave cant access its register when processing jack event. how about to get pm up when dispatching irq event to slave only ?  Thanks!

Reply:
@RanderWang I see your point but help me understand when the Slave device would be pm_runtime suspended when you get an interrupt. is this in the case where the master is active but the slave suspends? Is this in the case where you do clock stop with the DSP active? Is this in the case where you do clock stop and a complete re-enumeration.

I trust that this is real, but since I didn't do any experiment on my side I can only rely on your explanation to reconstruct the idea. Thanks!

Reply:
> @RanderWang I see your point but help me understand when the Slave device would be pm_runtime suspended when you get an interrupt. is this in the case where the master is active but the slave suspends? Is this in the case where you do clock stop with the DSP active? 
Is this in the case where you do clock stop and a complete re-enumeration.

if DSP is active, so device is in D0ix,  driver doesn't need a complete re-enumeration and clock stop can works as the flow defined in spec. If device is in D3 and we still need to enable clock stop, slave may send unattached or alert event (actually alert event is rare and it should be incorrect for slave lost sync with master, and I got it only on one device) to master. I got Io timeout when alert event was sent to master. This pm sync can fix this io timeout.

Reply:
sorry, not getting your points. There are multiple devices, the PCI, master and Slave ones.

There is also a difference between the wake in clock stop, and alerts that can only happen when the device is attached and with a non-zero Device Number.

If the PCI device is in D3, the wake from the Slave will result in a bus reset and re-enumeration.  The Slave will become re-attached.

So overall I just don't have any idea of when this IO error happens? Can you try to write a sequence (maybe with bullet points to that the transitions are clear), that may be more helpful that sentences with and/or statements.

Reply:
> sorry, not getting your points. There are multiple devices, the PCI, master and Slave ones.
> 
> There is also a difference between the wake in clock stop, and alerts that can only happen when the device is attached and with a non-zero Device Number.
> 
> If the PCI device is in D3, the wake from the Slave will result in a bus reset and re-enumeration. The Slave will become re-attached.
> 
> So overall I just don't have any idea of when this IO error happens? Can you try to write a sequence (maybe with bullet points to that the transitions are clear), that may be more helpful than sentences with and/or statements.

yes,  it is good idea to make it more clear
if the PCI device is in D3 and clock stop mode 0 is enabled
(1) plug a headphone, the alert event sent by slave is captured by glue hw, not master
(2) glue hw will wake up pci device and power on master, then initialize the master
(3) master enable clk and bus data to communicate with slave
(4) slave found the synchronization is not correct,  and send alert event or unattached event to master. here I don't know why slave send alert event, and at this time IO transfer timed out when accessing slave register (sdw_read(slave, SDW_SCP_INT1)). And I also checked slave status by cdns_readl(cdns, CDNS_MCP_SLAVE_STATE) when IO transfer timed out for alert event, the status showed slave was unattached!  So I think if master is powered off, we need to reset bus before communicating with slave.

Reply:
Something is very odd here.
I reproduced this error when the PCI device is active, the master is active and the slave device is suspended
That's not at all what your description is.
I took your code but I'd really like to understand this better.

Reply:
@plbossart ,  I mentioned in my comments: it is rare, and last week it also could be reproduced on one device at @bardliao. He tested my branch and also got such issue.

I could easily reproduced it  with the branch in this PR: in this branch, I didn't reset bus and clear slave status to unattached.

please check my comments and log. https://github.com/thesofproject/linux/pull/1498#issuecomment-557010741

please check line 307 in the log:  master got a alert event
https://github.com/thesofproject/linux/files/3874045/sdw_io_timeout.log